### PR TITLE
[CDAP-12091] add Define Variable and Set Counter directive

### PR DESCRIPTION
**Pull request by elfenheart**
JIRA: https://issues.cask.co/browse/CDAP-12091

### CODE REVIEW

**Comment by tonybach** [comment]
LGTM 👍

**Comment by ajainarayanan** [comment]
Talked to Lea and @elfenheart . Need to do two things,

1. Add summary above the Apply button
2. Need to confirm with @nitinmotgi about possible filter options for narrowing it down to a cell.

File: cdap-ui/app/cdap/components/DataPrep/Directives/DefineVariable/index.js [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    if (this.props.isOpen && this.calculateOffset) {
+      this.calculateOffset();
+    }
+  }
+
+  componentWillUnmount() {
+    Mousetrap.unbind('enter', this.applyDirective);
+  }
+
+  preventPropagation(e) {
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/DataPrep/Directives/DefineVariable/index.js]: Can import this from `services/helpers`

File: cdap-ui/app/cdap/components/DataPrep/Directives/SetCounter/index.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    this.calculateOffset = setPopoverOffset.bind(this, document.getElementById('set-counter-directive'));
+  }
+
+  componentDidUpdate() {
+    if (this.props.isOpen && this.calculateOffset) {
+      this.calculateOffset();
+    }
+  }
+
+  preventPropagation(e) {
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/DataPrep/Directives/SetCounter/index.js]: here too

File: cdap-ui/app/cdap/components/DataPrep/Directives/SetCounter/index.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  renderVariableName() {
+    return (
+      <div>
+        <label className="control-label">
+          {T.translate(`${PREFIX}.variableNameLabel`)}
+        </label>
+
+        <input
+          type="text"
+          className="form-control"
```
Comment (tonybach) [inline cdap-ui/app/cdap/components/DataPrep/Directives/SetCounter/index.js]: I think you also need `mousetrap` class here

**Review by tonybach — COMMENTED** [review]
Minor comments

**Comment by tonybach** [comment]
Code LGTM 👍

**Comment by bdmogal** [comment]
@elfenheart I saw this yesterday, and I think we should make some small messaging changes in the Define Variable flow to make it clearer.

1. At the top level, the flow should consist of two things:
  i. Variable Name
  ii. Variable Value
2. In the Variable Value section, have the cell selection (row selection and column selection) like you have right now.
3. Perhaps separate the Variable name and value sections by a horizontal rule.

**Review by bdmogal — APPROVED** [review]

**Pull request merged**